openembedded-core-mbl/scripts/lib/mbl-signing/signing/fip/fiptool.py

```diff
@@ -0,0 +1,389 @@
+        cmd_opts=opts,
+    )
+
+
+def remove(
+    img_spec: "ImageSpec",
+    fip_path: str,
+    align: int = 0,
+    blob_uuid: str = "",
+    force: bool = False,
+    out: pathlib.Path = None,
+) -> int:
+    """Remove images from the FIP.
+
+    :param ImageSpec img_spec: ImageSpec object describing images to remove.
+    :param str fip_filename: Path to an existing FIP image.
+    """
+    opts = _make_cmd_opts(
+        cmd="remove",
+        func=remove,
+        align=align,
+        blob_uuid=blob_uuid,
+        force=force,
+        out=out,
+    )
+    return _invoke(
+        "remove", img_spec=img_spec, fip_filename=fip_path, cmd_opts=opts
+    )
+
+
+def info(fip_path: str) -> "ImageSpec":
+    """List images contained in FIP.
+
+    :param str fip_filepath: Path to an existing FIP.
+    """
+    return ImageSpec.from_info(_invoke("info", fip_filename=fip_path).stdout)
+
+
+def unpack(
+    existing_fip_filepath: pathlib.Path,
+    img_spec: "ImageSpec" = None,
+    out: pathlib.Path = None,
+    blob_uuid: str = "",
+    blob_file: str = "",
+    force: bool = False,
+) -> "ImageSpec":
+    """Unpack images from FIP.
+
+    :param str existing_fip_filepath: Path to an existing FIP.
+    :param str out: Output path for images.
+    """
+    opts = _make_cmd_opts(
+        cmd="unpack",
+        func=unpack,
+        blob_uuid=blob_uuid,
+        blob_file=blob_file,
+        force=force,
+        out=out,
+    )
+    _invoke(
+        "unpack",
+        fip_filename=existing_fip_filepath,
+        img_spec=img_spec,
+        cmd_opts=opts,
+    )
+    return info(existing_fip_filepath)
+
+
+def _make_cmd_opts(cmd, func, *args, **kwargs):
+    return CommandOpts.from_boundargs(
+        cmd, inspect.signature(func).bind_partial(*args, **kwargs).arguments
+    )
+
+
+@fiptool_error_logger
+def _invoke(
+    cmd,
+    *args,
+    img_spec: "ImageSpec" = None,
+    fip_filename: pathlib.Path = pathlib.Path(),
+    cmd_opts: "CommandOpts" = None
+):
+    """Invoke the ATF fiptool with the specified options."""
+    fiptool_cmd = [ToolPaths.FIPTOOL_BIN, cmd, *args]
+    if img_spec is not None:
+        fiptool_cmd.append(str(img_spec))
+    if cmd_opts is not None:
+        fiptool_cmd.append(str(cmd_opts))
+    fiptool_cmd.append(str(fip_filename))
+    logger.debug(
+        "Invoking fiptool with the following command: {}".format(fiptool_cmd)
+    )
+    return subprocess.run(
+        fiptool_cmd, check=True, capture_output=True, text=True
+    )
+
+
+class OptsDict(UserDict):
+    """Dict which generates a flattened list of options when str is called.
+
+    This dict-like object is used to invoke the ATF fiptool with the
+    correct options.
+
+    When this object's str method is called it produces an option string
+    intended to be passed to the fiptool CLI.
+
+    __setitem__ has been overridden to check for 'valid opts' i.e valid keys.
+
+    This object is intended to be subclassed, the valid_opts attribute should
```

Comment:
maybe we need to define an __init__ which raises a NotImplementedError exception

Reply:
Instead I've marked the `__str__` method as an `@abstractmethod' which prevents the class being instantiated by raising a `TypeError` (and any attempted instantiation will be flagged in linters such as flake8/pylint).